### PR TITLE
fix(transformers): pattern add domain transformer - enable replace_existing

### PR DESCRIPTION
**Pull request by asikowitz**
Propagates the `replace_existing` config flag and adds tests for the pattern add domain transformer.

Note, as demonstrated by the tests, that if the pattern does not match and `replace_existing = True`, it still replaces the domains for that entity, just with `[]`. This is the same behavior as other transformers with the `replace_existing` flag and patterns, like the pattern add dataset ownership transformer.

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/tests/unit/test_transform_dataset.py [hunk 1]
```diff
@@ -1822,6 +1825,209 @@ def fake_get_domain(entity_urn: str) -> models.DomainsClass:
...
+
+    assert len(output) == 2
+    assert output[0] is not None
+    assert output[0].record is not None
+    assert isinstance(output[0].record, MetadataChangeProposalWrapper)
+    assert output[0].record.aspect is not None
+    assert isinstance(output[0].record.aspect, models.DomainsClass)
+    transformed_aspect = cast(models.DomainsClass, output[0].record.aspect)
+    assert len(transformed_aspect.domains) == 2
+    assert gslab_domain in transformed_aspect.domains
```
Comment (jjoyce0510) [inline metadata-ingestion/tests/unit/test_transform_dataset.py:1855]: So PATCH is the default? (if semantics are not specified)

File: metadata-ingestion/src/datahub/ingestion/transformer/dataset_domain.py [hunk 2]
```diff
@@ -155,6 +155,7 @@ def resolve_domain(domain_urn: str) -> DomainsClass:
         generic_config = AddDatasetDomainSemanticsConfig(
             get_domains_to_add=resolve_domain,
             semantics=config.semantics,
+            replace_existing=config.replace_existing,
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/transformer/dataset_domain.py:158]: Well that was simple :)

File: metadata-ingestion/tests/unit/test_transform_dataset.py (hunk 1 above)
Comment (asikowitz) [reply metadata-ingestion/tests/unit/test_transform_dataset.py:1855]: No, OVERWRITE is the default, but if you don't have `replace_existing = True` it appends new domains to the existing domains for the entity. https://datahubproject.io/docs/metadata-ingestion/docs/transformer/dataset_transformer#relationship-between-replace_existing-and-semantics explains all the possible situations, although I don't like how this is kinda confusing

Comment (jjoyce0510) [reply metadata-ingestion/tests/unit/test_transform_dataset.py:1855]: Ah this is quite confusing... 

So IF the aspect is a DomainsClass aspect, then an OVERWRITE will mean "we merge into this aspect".

If the aspect is not a DomainsClass aspect, then we'll just generate a new DomainsClass aspect altogether. 

I think this makes sense but the documentation is a bit hard to follow for sure.

Maybe another set of cases that would be useful to include in the tests are what happens when a non-domains aspect (e.g. DatasetPropertiesClass) is sent through the pipeline (to show how it behaves then)

Comment (asikowitz) [reply metadata-ingestion/tests/unit/test_transform_dataset.py:1855]: The way I see it, the semantics just says whether you edit the existing row (via a `AddDatasetDomain` aspect) or create a new one (via a `DomainsClass` aspect). `replace_existing` determines whether you append to the existing domains list, or truncate. And, confusingly, pattern matching does absolutely nothing to change this logic -- transformers are applied on all change event / proposals with matching `aspect_name`, the pattern matching just allows you to say which entities get which domain, and if there's no match then they get no domains.

Shouldn't the pipeline around transformers (i.e. `base_transformer.py::BaseTransformer.transform` and `_transform_or_record_mcp` in that same class) handle that filtering? I would expect testing around that to be on the base transformer, rather than every transformer than inherits from the base transformer.

Comment (jjoyce0510) [reply metadata-ingestion/tests/unit/test_transform_dataset.py:1855]: Hmm my understanding of this transformer is that it can handle ANY aspects not just incoming domain aspects produced by the Ingestion Source... Maybe that's a bad interpretation

Comment (jjoyce0510) [reply metadata-ingestion/tests/unit/test_transform_dataset.py:1855]: It seems that way. So for example, if I configure a Transformer with an Ingestion Source that does not itself sent DomainsClass aspects (some do not), then this transformer will intercept whatever aspect it _does_ produce (e.g. datasetProperties) and will inject the Domains aspect (this would be my expectation, at least)

Comment (jjoyce0510) [reply metadata-ingestion/tests/unit/test_transform_dataset.py:1877]: Got it - so yeah the parent class adds filtering on the Domains aspect... Very odd because the source should not need to produce a Domains aspect for this to apply... Let me see if my understanding of the docs aligns

**Review by jjoyce0510 — APPROVED** [review]
